File: app/src/main/java/com/letianpai/robot/control/broadcast/battery/ChargingUpdateCallback.java
package com.letianpai.robot.control.broadcast.battery;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by liujunbin
 */

public class ChargingUpdateCallback {
    public static final int LOW_BATTERY_SHUTDOWN_STANDARD = 20;
    public static final int LOW_BATTERY_NOTICE = 30;
    private int battery;
    private boolean isCharging;
    private List<ChargingUpdateListener> mChargingListenerList = new ArrayList();

    private static class ChargingUpdateCallBackHolder {
        private static ChargingUpdateCallback instance = new ChargingUpdateCallback();
    }

    private ChargingUpdateCallback() {

    }

    public static ChargingUpdateCallback getInstance() {
        return ChargingUpdateCallBackHolder.instance;
    }

    public interface ChargingUpdateListener {
        void onChargingUpdateReceived(boolean changingStatus, int percent);
        void onChargingUpdateReceived(boolean changingStatus, int percent,int chargePlug);

    }
    public void registerChargingStatusUpdateListener(ChargingUpdateListener listener) {
        if(mChargingListenerList != null){
            mChargingListenerList.add(listener);
        }
    }
    public void unregisterChargingStatusUpdateListener(ChargingUpdateListener listener) {
        if(mChargingListenerList != null){
            mChargingListenerList.remove(listener);
        }
    }

    public void setChargingStatus(boolean changingStatus,int percent) {
        this.battery = percent;
        this.isCharging = changingStatus;
        for (int i = 0;i<mChargingListenerList.size();i++){
            if(mChargingListenerList.get(i) != null){
                mChargingListenerList.get(i).onChargingUpdateReceived(changingStatus,percent);
            }
        }
    }

    public void setChargingStatus(boolean changingStatus,int percent,int chargePlug) {
        this.battery = percent;
        this.isCharging = changingStatus;
        for (int i = 0;i<mChargingListenerList.size();i++){
            if(mChargingListenerList.get(i) != null){
                mChargingListenerList.get(i).onChargingUpdateReceived(changingStatus,percent,chargePlug);
            }
        }
    }

    public int getBattery() {
        return battery;
    }

    public boolean isCharging() {
        return isCharging;
    }
}

File: app/src/main/java/com/letianpai/robot/control/broadcast/battery/ChargingUpdateCallback.kt
package com.letianpai.robot.control.broadcast.battery

/**
 * Created by liujunbin
 */
class ChargingUpdateCallback private constructor() {
    var battery: Int = 0
        private set
    var isCharging: Boolean = false
        private set
    private val mChargingListenerList: MutableList<ChargingUpdateListener?> = ArrayList()

    private object ChargingUpdateCallBackHolder {
        val instance: ChargingUpdateCallback = ChargingUpdateCallback()
    }

    interface ChargingUpdateListener {
        fun onChargingUpdateReceived(changingStatus: Boolean, percent: Int)
        fun onChargingUpdateReceived(changingStatus: Boolean, percent: Int, chargePlug: Int)
    }

    fun registerChargingStatusUpdateListener(listener: ChargingUpdateListener?) {
        if (mChargingListenerList != null) {
            mChargingListenerList.add(listener)
        }
    }

    fun unregisterChargingStatusUpdateListener(listener: ChargingUpdateListener?) {
        if (mChargingListenerList != null) {
            mChargingListenerList.remove(listener)
        }
    }

    fun setChargingStatus(changingStatus: Boolean, percent: Int) {
        this.battery = percent
        this.isCharging = changingStatus
        for (i in mChargingListenerList.indices) {
            if (mChargingListenerList[i] != null) {
                mChargingListenerList[i]!!.onChargingUpdateReceived(changingStatus, percent)
            }
        }
    }

    fun setChargingStatus(changingStatus: Boolean, percent: Int, chargePlug: Int) {
        this.battery = percent
        this.isCharging = changingStatus
        for (i in mChargingListenerList.indices) {
            if (mChargingListenerList[i] != null) {
                mChargingListenerList[i]!!
                    .onChargingUpdateReceived(changingStatus, percent, chargePlug)
            }
        }
    }

    companion object {
        const val LOW_BATTERY_SHUTDOWN_STANDARD: Int = 20
        const val LOW_BATTERY_NOTICE: Int = 30
        val instance: ChargingUpdateCallback
            get() = ChargingUpdateCallBackHolder.instance
    }
}
